sql/field.cc

```diff
@@ -6324,8 +6324,10 @@ uchar *Field_string::pack(uchar *to, const uchar *from, uint max_length,
            (this is for not packing padding adding bytes in BINARY
            fields).
   */
-  if (field_charset->mbmaxlen == 1) {
-    while (length && from[length - 1] == field_charset->pad_char) length--;
+  if (key_type() != HA_KEYTYPE_BINARY) {
```

Comment:
I don't think the body of this if statement is correct. It should both the `mbmaxlen == 1` case and the else statement.

You can see that the re-recorded results are actually incorrect, because we're spacing padding varchars. This is because we're now keeping padding characters for charsets with mbmaxlen  > 1 (eg. utf8mb4)